tests/validation/test_transaction_structure.py

```diff
@@ -0,0 +1,159 @@
+"""
+All tests of transaction structure. The concern here is that transaction
+structural / schematic issues are caught when reading a transaction
+(ie going from dict -> transaction).
+"""
+
+import pytest
+
+from bigchaindb.common.exceptions import (AmountError, InvalidHash,
+                                          SchemaValidationError)
+from bigchaindb.models import Transaction
+
+
+################################################################################
+# Helper functions
+
+
+def validate(tx):
+    if isinstance(tx, Transaction):
+        tx = tx.to_dict()
+    Transaction.from_dict(tx)
+
+
+def validate_raises(tx, exc=SchemaValidationError):
+    with pytest.raises(exc):
+        validate(tx)
+
+
+# We should test that validation works when we expect it to
+def test_validation_passes(create_tx):
+    validate(create_tx)
+
+
+################################################################################
+# ID
+
+
+def test_tx_serialization_hash_function(create_tx):
+    import sha3
+    import json
+    tx = create_tx.to_dict()
+    tx['inputs'][0]['fulfillment'] = None
+    del tx['id']
+    payload = json.dumps(tx, skipkeys=False, sort_keys=True,
+                         separators=(',', ':'))
+    assert sha3.sha3_256(payload.encode()).hexdigest() == create_tx.id
+
+
+def test_tx_serialization_with_incorrect_hash(create_tx):
+    tx = create_tx.to_dict()
+    tx['id'] = 'a' * 64
+    validate_raises(tx, InvalidHash)
+
+
+################################################################################
+# Operation
+
+def test_validate_invalid_operation(create_tx):
+    create_tx.operation = 'something invalid'
+    validate_raises(create_tx)
+
+
+################################################################################
+# Metadata
+
+def test_validate_fails_metadata_empty_dict(create_tx):
+    create_tx.metadata = {'a': 1}
+    validate(create_tx)
+    create_tx.metadata = None
+    validate(create_tx)
+    create_tx.metadata = {}
+    validate_raises(create_tx)
+
+
+################################################################################
+# Asset
+
+def test_transfer_asset_schema(signed_transfer_tx):
+    tx = signed_transfer_tx.to_dict()
+    validate(tx)
+    tx['asset']['data'] = {}
+    validate_raises(tx)
+    del tx['asset']['data']
+    tx['asset']['id'] = 'b' * 63
+    validate_raises(tx)
```

Comment:
Can the asset have no `data` field?
Is this raising because asset has no `data` field or because it has a malformed hash?